accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,671 @@
+        - User-defined classes
+        - Boxed value types
+        - Delegates
+
+## Overidden members for a modified type
+Below are the non-trivial behaviors overridden by a modified type.
+
+### `bool IsModifiedType { get; }`
+Returns `true`.
+
+### `Type UnderlyingSystemType { get; }`
+The underlying type, such as `typeof(Int32)`.
+
+### `ToString()`
+Returns `UnderlyingSystemType.ToString()`. We could return a string representation of the underlying type plus its custom modifiers, but that is not considered a high enough priority at this time and the grammar would be somewhat arbitrary.
+
+### `Name`, `FullName`, `AssemblyQualifiedName`, `Assembly` and `Namespace` properties
+A modified type is an internal class called "System.RuntimeModifiedType" class that derives from a "System.ModifiedType" which derives from "System.Type". So these methods just return the expected values there.
+
+### `GetFunctionPointerParameters()`
+This returns modified types for each function pointer parameter which allows the user to call `GetRequiredCustomModifiers()` or `GetOptionalCustomModifiers()`.
+
+This is listed here since this method also works on unmodified types, although unmodified types don't allow `GetRequiredCustomModifiers()` or `GetOptionalCustomModifiers()` to be called (an `InvalidOperationException` is thrown).
+
+# Proposed APIs
+## System.Type
+```diff
+namespace System
+{
+    public abstract class Type
+    {
+        // Throws InvalidOperationException if IsFunctionPointer = false.
++       public virtual Type GetFunctionPointerReturnType();
++       public virtual Type[] GetFunctionPointerParameterTypes();
++       public virtual bool IsFunctionPointer { get; }
+
++       public virtual bool IsUnmanagedFunctionPointer { get; }
+
+        // Returns whether the type is a modified type.
+        // A modified type is either:
+        //  - A function pointer obtained though FieldInfo\PropertyInfo\ParametersInfo.GetModified*Type().
+        //  - A function pointer parameter obtained by 'GetFunctionPointerParameterTypes()[index]' from a modified function pointer type.
++       public virtual bool IsModifiedType { get; }
+
+        // Updated semantics: if IsModifiedType = true, returns the unmodified type such as `typeof(Int32)`.
+        public virtual bool UnderlyingSystemType { get; }
+
+        // These only work if IsModifiedType = true. InvalidOperationException otherwise.
++       public virtual Type[] GetRequiredCustomModifiers();
++       public virtual Type[] GetOptionalCustomModifiers();
+        // Throws InvalidOperationException if UnderlyingSystemType.IsFunctionPointer = false
++       public virtual Type[] GetFunctionPointerCallingConventions(); 
+
+        // This is used with 'Type.GetMethod(..., Type[] types)' to find a method with parameters that
+        // are function pointers. For a similar reference, see Type.MakePointerType().
++       public static Type MakeFunctionPointerType(Type[] functionPointerArguments);
```

Comment:
I'd suggest separating the return type - it's a trial and error to remember whether return type comes first or last in schemes where it's a single array.

Do we also need a bool to signify if this is an unmanaged/managed function pointer?

(My reading of this is that `MakeFunctionPointerType` would return a type that has a valid `TypeHandle` and these would be the only constituents that form the runtime identity .)

Reply:
Thanks. I updated the design for separate parameters + isUnmanaged but based on other feedback, moved this to the "future" section. The work-around is not so bad, plus likely to be very rare anyway unlike the original support which was added for looking up generic methods and thus much more likely to be used.